sled-agent/src/support_bundle/logs.rs

```diff
@@ -0,0 +1,253 @@
+use std::collections::BTreeSet;
+
+use camino::Utf8PathBuf;
+use dropshot::{Body, HttpError};
+use range_requests::PotentialRange;
+use sled_diagnostics::{SledDiagnosticsLogError, SledDiagnosticsLogs};
+use slog::Logger;
+use slog_error_chain::InlineErrorChain;
+use tokio::io::{AsyncReadExt, AsyncSeekExt};
+use tokio_util::io::ReaderStream;
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Failed to join tokio task: {0}")]
+    Join(#[from] tokio::task::JoinError),
+
+    #[error(transparent)]
+    Logs(#[from] SledDiagnosticsLogError),
+
+    #[error("Path \"{0}\" not found")]
+    NotFound(Utf8PathBuf),
+
+    #[error("Directory \"{0}\" is not a valid log path")]
+    InvalidLogDir(Utf8PathBuf),
+
+    #[error("Path \"{0}\" should not be the final component")]
+    MissingFilename(Utf8PathBuf),
+
+    #[error(transparent)]
+    Io(#[from] std::io::Error),
+
+    #[error(transparent)]
+    Range(#[from] range_requests::Error),
+
+    #[error("Not a file")]
+    NotAFile,
+}
+
+fn err_str(err: &dyn std::error::Error) -> String {
+    InlineErrorChain::new(err).to_string()
+}
+
+impl From<Error> for HttpError {
+    fn from(err: Error) -> Self {
+        match err {
+            Error::InvalidLogDir(_) => {
+                HttpError::for_bad_request(None, "Invalid log dir".to_string())
+            }
+            Error::NotAFile => {
+                HttpError::for_bad_request(None, "Not a file".to_string())
+            }
+            Error::NotFound(path) => HttpError::for_not_found(
+                None,
+                format!("Log file {path} not found"),
+            ),
+            err => HttpError::for_internal_error(err_str(&err)),
+        }
+    }
+}
+
+pub struct SupportBundleLogs<'a> {
+    log: &'a Logger,
+}
+
+impl<'a> SupportBundleLogs<'a> {
+    pub fn new(log: &'a Logger) -> Self {
+        Self { log }
+    }
+
+    pub async fn get_index(&self) -> Result<SledDiagnosticsLogs, Error> {
+        tokio::task::spawn_blocking(move || sled_diagnostics::find_logs())
+            .await
+            .map_err(Error::Join)?
+            .map_err(Error::Logs)
+    }
+
+    /// Get a log file from disk preforming normalization and validation along
+    /// the way.
+    pub async fn get_log_file(
+        &self,
+        req: &str,
+        range: Option<PotentialRange>,
+    ) -> Result<http::Response<Body>, Error> {
+        let valid_dirs =
+            tokio::task::spawn_blocking(|| sled_diagnostics::valid_log_dirs())
+                .await
+                .map_err(Error::Join)?
+                .map_err(Error::Logs)?;
```

Comment:
minor: I'm curious about the performance impact here. Accessing each individual log file requires constructing this directory of all log directories - not a bad choice for a first implementation, but I'm curious how costly that is relative to reading out the logs themselves.

**We don't have to do this yet** but as one example, we could cache this result, and refresh it + retry if `validate_log_dir` returns an error.

Reply:
I will grab some data here again but iirc grabbing the directories themselves is relatively fast compared to retrieving all of the log files themselves.

We had discussed earlier keeping this directory in memory and refreshing it but for whatever reason we decided that wasn't a good idea - I would be happy to to implement such a cache in the get_log_file code path.